File: appmonet-sdk/appmonet-sdk-mopub/src/main/java/com/monet/bidder/AppMonetStaticNativeAd.java
package com.monet.bidder;

import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.mopub.nativeads.BaseNativeAd;
import com.mopub.nativeads.ClickInterface;
import com.mopub.nativeads.ImpressionInterface;
import com.mopub.nativeads.ImpressionTracker;
import com.mopub.nativeads.NativeClickHandler;
import com.mopub.nativeads.NativeErrorCode;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * Class which holds the logic and interface callbacks for the native ad to be displayed and
 * interacted with.
 */
public class AppMonetStaticNativeAd extends BaseNativeAd implements ClickInterface, ImpressionInterface {
  private static final MonetLogger logger = new MonetLogger("AppMonetStaticNativeAd");
  private final CustomEventNative.CustomEventNativeListener mCustomEventNativeListener;
  private final HashMap<String, Object> mExtras;
  private final AppMonetNativeEventCallback appMonetNativeEventCallback;

  enum Parameter {
    IMPRESSION_TRACKER("imptracker", false),
    TITLE("title", false),
    TEXT("text", false),
    ICON("icon", false),
    CALL_TO_ACTION("ctatext", false);

    @NonNull
    final String name;
    final boolean required;

    Parameter(@NonNull final String name, final boolean required) {
      this.name = name;
      this.required = required;
    }

    @Nullable
    static Parameter from(@NonNull final String name) {
      for (final Parameter parameter : values()) {
        if (parameter.name.equals(name)) {
          return parameter;
        }
      }
      return null;
    }

    @NonNull
    private static final Set<String> requiredKeys = new HashSet<>();

    static {
      for (final Parameter parameter : values()) {
        if (parameter.required) {
          requiredKeys.add(parameter.name);
        }
      }
    }
  }

  @NonNull
  private final Map<String, String> serverExtras;
  @Nullable
  private View mainView;
  @Nullable
  private String mTitle;
  @Nullable
  private String mIcon;
  @Nullable
  private String mText;
  @Nullable
  private String mCallToAction;
  @Nullable
  private View media;
  private boolean mImpressionRecorded;
  private int mImpressionMinTimeViewed = 1000;
  @NonNull
  private final ImpressionTracker mImpressionTracker;
  @NonNull
  private final NativeClickHandler mNativeClickHandler;

  public AppMonetStaticNativeAd(@NonNull Map<String, String> serverExtras, @Nullable View view, @NonNull ImpressionTracker impressionTracker,
                                @NonNull NativeClickHandler nativeClickHandler, @NonNull CustomEventNative.CustomEventNativeListener customEventNativeListener,
                                @NonNull AppMonetNativeEventCallback appMonetNativeEventCallback) {
    this.media = view;
    this.mCustomEventNativeListener = customEventNativeListener;
    this.serverExtras = serverExtras;
    this.mImpressionTracker = impressionTracker;
    this.mNativeClickHandler = nativeClickHandler;
    this.appMonetNativeEventCallback = appMonetNativeEventCallback;
    mExtras = new HashMap<>();
  }

  @Nullable
  public View getMainView() {
    return mainView;
  }

  public void setMainView(@Nullable View mainView) {
    this.mainView = mainView;
  }

  @Nullable
  public String getTitle() {
    return mTitle;
  }

  @Nullable
  public String getIcon() {
    return mIcon;
  }

  public void setTitle(@Nullable String title) {
    this.mTitle = title;
  }

  public void setIcon(@Nullable String iconSrc) {
    this.mIcon = iconSrc;
  }

  @Nullable
  public String getText() {
    return mText;
  }

  public void setText(@Nullable String text) {
    this.mText = text;
  }

  @Nullable
  public String getCallToAction() {
    return mCallToAction;
  }

  public void setCallToAction(@Nullable String callToAction) {
    this.mCallToAction = callToAction;
  }

  @Nullable
  public View getMedia() {
    return media;
  }

  public void setMedia(@Nullable View media) {
    this.media = media;
  }

  @Override
  public void prepare(@NonNull View view) {
    this.mImpressionTracker.addView(view, this);
    this.mNativeClickHandler.setOnClickListener(view, this);
  }

  @Override
  public void clear(@NonNull View view) {
    this.mImpressionTracker.removeView(view);
    this.mNativeClickHandler.clearOnClickListener(view);
  }

  @Override
  public void destroy() {
    this.mImpressionTracker.destroy();
    if (media != null) {
      this.appMonetNativeEventCallback.destroy(media);
    }
  }

  @Override
  public void recordImpression(@NonNull View view) {
    notifyAdImpressed();
  }

  @Override
  public final int getImpressionMinPercentageViewed() {
    return 50;
  }

  @Override
  public Integer getImpressionMinVisiblePx() {
    return null;
  }

  @Override
  public final int getImpressionMinTimeViewed() {
    return this.mImpressionMinTimeViewed;
  }

  @Override
  public final boolean isImpressionRecorded() {
    return this.mImpressionRecorded;
  }

  @Override
  public final void setImpressionRecorded() {
    this.mImpressionRecorded = true;
  }

  //
  @Override
  public void handleClick(@NonNull View view) {
    if (media != null) {
      appMonetNativeEventCallback.onClick(media);
      this.notifyAdClicked();
    }
  }

  @NonNull
  final public Map<String, Object> getExtras() {
    return new HashMap<>(mExtras);
  }

  public void onAdClicked() {
    notifyAdClicked();
  }

  public void loadAd() {
    if (!containsRequiredKeys(serverExtras)) {
      mCustomEventNativeListener.onNativeAdFailed(NativeErrorCode.SERVER_ERROR_RESPONSE_CODE);
    }

    Set<String> keys = serverExtras.keySet();
    for (String key : keys) {
      final Parameter parameter = Parameter.from(key);
      if (parameter != null) {
        try {
          addInstanceVariable(parameter, serverExtras.get(key));
        } catch (ClassCastException e) {
          mCustomEventNativeListener.onNativeAdFailed(NativeErrorCode.UNEXPECTED_RESPONSE_CODE);
        }
      } else {
        addExtra(key, serverExtras.get(key));
      }
    }

    mCustomEventNativeListener.onNativeAdLoaded(this);
  }

  private boolean containsRequiredKeys(@NonNull final Map<String, String> serverExtras) {
    final Set<String> keys = serverExtras.keySet();
    return keys.containsAll(Parameter.requiredKeys);
  }

  private void addInstanceVariable(@NonNull final Parameter key,
                                   @Nullable final Object value) {
    try {
      switch (key) {
        case ICON:
          setIcon((String) value);
          break;
        case CALL_TO_ACTION:
          setCallToAction((String) value);
          break;
        case TITLE:
          setTitle((String) value);
          break;
        case TEXT:
          setText((String) value);
          break;
        default:
          logger.debug("Unable to add JSON key to internal mapping: " + key.name);
          break;
      }
    } catch (ClassCastException e) {
      if (!key.required) {
        logger.debug("Ignoring class cast exception for optional key: " + key.name);
      } else {
        throw e;
      }
    }
  }

  final void addExtra(@NonNull final String key, @Nullable final Object value) {
    mExtras.put(key, value);
  }

  void swapViews(AppMonetViewLayout view, AdServerBannerListener listener) {
    ((AppMonetViewLayout) this.media).swapViews(view, listener);
  }
}

File: appmonet-sdk/appmonet-sdk-mopub/src/main/java/com/monet/bidder/AppMonetStaticNativeAd.kt
package com.monet.bidder

import android.view.View
import com.monet.AdServerBannerListener
import com.monet.bidder.AppMonetStaticNativeAd.Parameter.CALL_TO_ACTION
import com.monet.bidder.AppMonetStaticNativeAd.Parameter.ICON
import com.monet.bidder.AppMonetStaticNativeAd.Parameter.TEXT
import com.monet.bidder.AppMonetStaticNativeAd.Parameter.TITLE
import com.mopub.nativeads.BaseNativeAd
import com.mopub.nativeads.ClickInterface
import com.mopub.nativeads.CustomEventNative.CustomEventNativeListener
import com.mopub.nativeads.ImpressionInterface
import com.mopub.nativeads.ImpressionTracker
import com.mopub.nativeads.NativeClickHandler
import com.mopub.nativeads.NativeErrorCode.SERVER_ERROR_RESPONSE_CODE
import com.mopub.nativeads.NativeErrorCode.UNEXPECTED_RESPONSE_CODE
import java.util.HashMap
import java.util.HashSet

/**
 * Class which holds the logic and interface callbacks for the native ad to be displayed and
 * interacted with.
 */
class AppMonetStaticNativeAd(
  private val serverExtras: Map<String, String>,
  var media: View?,
  private val mImpressionTracker: ImpressionTracker,
  private val mNativeClickHandler: NativeClickHandler,
  private val mCustomEventNativeListener: CustomEventNativeListener,
  private val appMonetNativeEventCallback: AppMonetNativeEventCallback
) : BaseNativeAd(), ClickInterface, ImpressionInterface {
  private val mExtras: HashMap<String, Any?> = HashMap()

  internal enum class Parameter(
    val type: String,
    val required: Boolean
  ) {

    IMPRESSION_TRACKER("imptracker", false),
    TITLE("title", false),
    TEXT("text", false),
    ICON("icon", false),
    CALL_TO_ACTION("ctatext", false);

    companion object {
      fun from(name: String): Parameter? {
        for (parameter in values()) {
          if (parameter.type == name) {
            return parameter
          }
        }
        return null
      }

      val requiredKeys: MutableSet<String> = HashSet()

      init {

        for (parameter in values()) {
          if (parameter.required) {
            requiredKeys.add(parameter.type)
          }
        }
      }
    }
  }

  var mainView: View? = null
  var title: String? = null
  var icon: String? = null
  var text: String? = null
  var callToAction: String? = null
  private var mImpressionRecorded = false
  private val mImpressionMinTimeViewed = 1000
  override fun prepare(view: View) {
    mImpressionTracker.addView(view, this)
    mNativeClickHandler.setOnClickListener(view, this)
  }

  override fun clear(view: View) {
    mImpressionTracker.removeView(view)
    mNativeClickHandler.clearOnClickListener(view)
  }

  override fun destroy() {
    mImpressionTracker.destroy()
    if (media != null) {
      appMonetNativeEventCallback.destroy(media)
    }
  }

  override fun recordImpression(view: View) {
    notifyAdImpressed()
  }

  override fun getImpressionMinPercentageViewed(): Int {
    return 50
  }

  override fun getImpressionMinVisiblePx(): Int? {
    return null
  }

  override fun getImpressionMinTimeViewed(): Int {
    return mImpressionMinTimeViewed
  }

  override fun isImpressionRecorded(): Boolean {
    return mImpressionRecorded
  }

  override fun setImpressionRecorded() {
    mImpressionRecorded = true
  }

  //
  override fun handleClick(view: View) {
    if (media != null) {
      appMonetNativeEventCallback.onClick(media)
      notifyAdClicked()
    }
  }

  val extras: Map<String, Any?>
    get() = HashMap(mExtras)

  fun onAdClicked() {
    notifyAdClicked()
  }

  fun loadAd() {
    if (!containsRequiredKeys(serverExtras)) {
      mCustomEventNativeListener.onNativeAdFailed(SERVER_ERROR_RESPONSE_CODE)
    }
    val keys = serverExtras.keys
    for (key in keys) {
      val parameter = Parameter.from(key)
      if (parameter != null) {
        try {
          addInstanceVariable(parameter, serverExtras[key])
        } catch (e: ClassCastException) {
          mCustomEventNativeListener.onNativeAdFailed(UNEXPECTED_RESPONSE_CODE)
        }
      } else {
        addExtra(key, serverExtras[key])
      }
    }
    mCustomEventNativeListener.onNativeAdLoaded(this)
  }

  private fun containsRequiredKeys(serverExtras: Map<String, String>): Boolean {
    val keys = serverExtras.keys
    return keys.containsAll(Parameter.requiredKeys)
  }

  private fun addInstanceVariable(
    key: Parameter,
    value: Any?
  ) {
    try {
      when (key) {
        ICON -> icon = value as String?
        CALL_TO_ACTION -> callToAction = value as String?
        TITLE -> title = value as String?
        TEXT -> text = value as String?
        else -> logger.debug("Unable to add JSON key to internal mapping: " + key.type)
      }
    } catch (e: ClassCastException) {
      if (!key.required) {
        logger.debug("Ignoring class cast exception for optional key: " + key.type)
      } else {
        throw e
      }
    }
  }

  fun addExtra(
    key: String,
    value: Any?
  ) {
    mExtras[key] = value
  }

  fun swapViews(
    view: AppMonetViewLayout,
    listener: AdServerBannerListener<View?>
  ) {
    (media as AppMonetViewLayout?)!!.swapViews(view, listener)
  }

  companion object {
    private val logger = Logger("AppMonetStaticNativeAd")
  }

}